src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -2312,7 +2314,9 @@ def _ensure_service_principal(cli_ctx,
         else:
             # Nothing to load, make one.
             if not client_secret:
-                client_secret = binascii.b2a_hex(os.urandom(10)).decode('utf-8')
+                special_chars = '!#$%&*-+_.:;<>=?@][^}{|~)('
```

Comment:
Since these three lines are duplicated, maybe make this into a `_create_client_secret()` function?

Reply:
I am open to making it into a function. However, right now the duplication is across the acs and the aks functions for ensuring sp. Those two functions have an incredible amount of overlap but I assumed there was a logical reason for keeping duplicated code. My guess (not validated) is that since we are retiring acs in 2020 we want to be able to easily remove that code.

Either way, I am open to making it a function but eventually it might only have only one caller

Reply:
I think the real reason for so much duplication is that we haven't paid attention to where we could do these kind of refactorings, we've just added AKS features and not looked back. ACS is deprecated but there's no reason we can't share client code with AKS (at least for another ten months).

But it's just a suggestion—the PR looks good to me either way.

Reply:
Ahh makes sense. I think it is a good idea - it is now fixed